Npm

brettz9 opened the pull request:
- breaking change: Bump `engines` to avoid EOL versions (Node >=10) which fail with latest dev updates
- update: Bump Travis versions to 12 and 14, dropping EOL'd 8
- linting: apply latest prettier (adding prettierignore to avoid applying to test sources) and lint, avoiding duplicated literals in `generate-rules-sonar-meta`
- linting: apply latest `eslint-plugin-import`
- chore: add `.ncurc.js` to prevent `peerDependencies` auto-updating when running `ncu` (or `npm-check-updates`)
- chore(lint-staged): remove redundant `git add` (as recommended)
- npm: - Use repository URL (accepted in npm, package.json linter doesn't complain, and allows IDEs to auto-open in browser (e.g., on cmd-O)
- npm: Add ESLint 7 to peerDeps.
- npm: Update devDeps.

Re: lint-staged change, per https://github.com/okonet/lint-staged#v10

> If your task previously contained a git add step, please remove this. The automatic behaviour ensures there are less race-conditions, since trying to run multiple git operations at the same time usually results in an error.

The `.eslintignore` change is required by your ruling given the update to ESLint 7 (which has different ignore file defaults from ESlint 6).

brettz9 commented:
I've updated such that the build is now passing.

vilchik-elena commented:
Uf, sorry I thought I commented on this PR yesterday, but looks like Github was down for some time.

It's not good idea to have such different changes in the same PR. 

Please create different PRs for the changes you have in mind. Thanks!

brettz9 commented:
If you don't mind, I'll keep this open until #158, #160, and #161 may be reviewed, and then amend this PR on top of any resulting changes, so this PR is indeed focused just on npm changes.

brettz9 commented:
While there are still a few minor tweaks in here I could separate out into separate PRs, most of the changes remaining are in fact related to npm updates, such as changes needed by the ESLint update or TypeScript updates. Currently, the only package not updated to the latest version in this PR is `eslint-plugin-import` which seems to have some issue or other.

brettz9 commented:
Ok, I've now stripped out everything which is not npm-related.

Note that everything in this PR is required as a result of updating devDependencies except for two small changes (though they are also npm related).

- chore: add `.ncurc.js` to prevent peerDependencies auto-updating when running ncu (or npm-check-updates)
- npm: - Use repository URL (accepted in npm, package.json linter doesn't complain, and allows IDEs to auto-open in browser (e.g., on cmd-O)

If you were open to it, it'd also be nice to specify an `author` and `contributors` field as suggested by `package.json` linter, not only to suppress the distracting warning but also for npmjs.com to list.

brettz9 commented:
I also just now added this change (as per the Travis version upgrade):

- Testing: Ensure test-ci script runs (now that Node 8 is removed)

And also added this comment correction:

- Docs: Avoid comment about "nodejs 6" being cause of failure; the need for `jest-sonar-reporter` is due to it not being in devDeps.

silverwind commented on package.json:
```diff
@@ -24,53 +28,70 @@
+      "prettier --write"
+    ]
+  },
+  "dependencies": {}
```
```suggestion
```

brettz9 replied:
Thanks for the suggestion. This actually serves two purposes though. It shuts up my package.json linter, avoiding the distraction of its messages. :-)  Of possibly more interest, with `package.json` files allowing these keys to be placed anywhere, it can be helpful to quickly search a file for `dependencies` to find them regardless of location. When one is in fast mode and/or jumping across projects, that can save a small bit of time and effort..

silverwind replied:
Okay, no problem. I thought it was an leftover of a `npm uninstall` which does not get rid of the key either.

silverwind commented:
Looking forward to this. Missed it earlier because of that title 😉.

saberduck commented:
thank you for this contribution! However, I agree with @vilchik-elena , it will be easier to review the changes if they are split into separate PR. From the quick look, I would separate the following changes into another PRs

- prettier update 
- dropping of NodeJS 8
- update of dependencies unrelated to ESLint 7
- ESLint 7

brettz9 commented:
If I take the time to break this up even further (this PR had actually been larger, and most had been related to updating to keep your dependencies current), can someone at least try to provide some reviews fairly quickly? I know you surely have other things on your plate, but it can be rather frustrating to have all of this back-and-forth stretched out over time.

brettz9 commented:
The reason for part of this frustration is that many issues are interconnected, and takes time to tease them out--e.g., an updated dependency may require an engines/Travis update, etc.

brettz9 commented:
There are a lot of intersecting pieces, so even teasing things out, to the extent it is possible, is not going to be easy to the extent even possible.

1. The dropping of Node 8 can be confirmed as a need for various changes, such as to the likes of ESLint 7 (or updating `lint-staged` with Node 8 dropped out of necessity for it in #171 ), all of which require Node 10+.
2. Updating prettier conflicts with `eslint-plugin-import`, requiring a rule to be disabled, and even with this there are now TS issues that were resolved when some combination of the other dep. updates was made (#172 ).
3. Updating the non-ESLint dependencies alone causes the TypeScript tests to fail.

If you can review the two passing PRs I submitted (#170 and #171 ), now passing, I can see about getting my other updates working. But if it takes a very long time to get these reviewed, I am frankly less confident my time will be well spent working to help get your project up to date.

If that works, I can then see about splitting ESLint 7 from the other dep. updates, but again, it may be difficult or impossible to tease out given all the TypeScript requirements and intersecting requirements.

saberduck commented:
@brettz9 I understand the frustration, and I am sorry but we currently don't have capacity to review such PRs in a timely manner. Especially when you are doing big changes that require thorough testing. I will try to handle the PRs you mentioned but please keep changes to the minimum, adding new tools, which are not strictly necessary is creating a lot of load on our side (we need to maintain this project in the long term, so we prefer to keep dependencies to the minimum).

silverwind commented:
I think focus should be on the eslint 7 peerDependency update. Is there a PR for it yet?

brettz9 commented:
Once #171 may be merged (or before if you like since this builds on top of that now), here is a very easy way to confirm that the changes in this PR are necessary (and cannot be done much more incrementally).

1. Use only my modifications for `package.json`. Many of these have to be upgraded together. (If you don't believe me, try updating one or two yourself alone and run the scripts; all power to you if you can, but I don't think you will find that any significant number can be updated independently, esp. the eslint, prettier, and typescript-related updates which are the bulk of the changes.)
1. Try running npm scripts and see that they fail.
1. Add my changes for these files (as the need for these may be harder to discover on your own, despite being small changes):
    1. `.eslintignore`
    1. `.prettierignore`
    1. `tsconfig.json`
1. Try running the scripts again and fixing what does not get auto-fixed (e.g., adding `output` to tests as eslint will insist and as I have done). You should see that most of the changes in my PR get added when running the new `prettier`.
1. Confirm that your now functioning local copy is the same as my PR.

etiennetremel commented:
What's the status of this PR, any timeline on when this will get reviewed?

brettz9 commented:
@saberduck : Now that I went ahead with the eslint-plugin-radar fork, should I just close these PRs, or do you think there is a chance you may revisit the project in the future?

saberduck commented:
You can close, they still remain available
